### PR TITLE
Add magic $_path stash key to docs tests

### PR DESCRIPTION
**Pull request by nik9000**
Adds a "magic" key to the yaml testing stash mostly for use with
documentation tests. When unstashing an object, `$_path` is the
path into the current position in the object you are unstashing.
This means that in docs tests you can use
`// TESTRESPONSEs/somevalue/$body.${_path}/` to mean "replace
`somevalue` with whatever is the response in the same position."

Compare how you must carefully mock out all the numbers in the profile
response without this change:
```
// TESTRESPONSE[s/"id": "\[2aE02wS1R8q_QFnYu6vDVQ\]\[twitter\]\[1\]"/"id": $body.profile.shards.0.id/]
// TESTRESPONSE[s/"rewrite_time": 51443/"rewrite_time": $body.profile.shards.0.searches.0.rewrite_time/]
// TESTRESPONSE[s/"score": 51306/"score": $body.profile.shards.0.searches.0.query.0.breakdown.score/]
// TESTRESPONSE[s/"time_in_nanos": "1873811"/"time_in_nanos": $body.profile.shards.0.searches.0.query.0.time_in_nanos/]
// TESTRESPONSE[s/"build_scorer": 2935582/"build_scorer": $body.profile.shards.0.searches.0.query.0.breakdown.build_scorer/]
// TESTRESPONSE[s/"create_weight": 919297/"create_weight": $body.profile.shards.0.searches.0.query.0.breakdown.create_weight/]
// TESTRESPONSE[s/"next_doc": 53876/"next_doc": $body.profile.shards.0.searches.0.query.0.breakdown.next_doc/]
// TESTRESPONSE[s/"time_in_nanos": "391943"/"time_in_nanos": $body.profile.shards.0.searches.0.query.0.children.0.time_in_nanos/]
// TESTRESPONSE[s/"score": 28776/"score": $body.profile.shards.0.searches.0.query.0.children.0.breakdown.score/]
// TESTRESPONSE[s/"build_scorer": 784451/"build_scorer": $body.profile.shards.0.searches.0.query.0.children.0.breakdown.build_scorer/]
// TESTRESPONSE[s/"create_weight": 1669564/"create_weight": $body.profile.shards.0.searches.0.query.0.children.0.breakdown.create_weight/]
// TESTRESPONSE[s/"next_doc": 10111/"next_doc": $body.profile.shards.0.searches.0.query.0.children.0.breakdown.next_doc/]
// TESTRESPONSE[s/"time_in_nanos": "210682"/"time_in_nanos": $body.profile.shards.0.searches.0.query.0.children.1.time_in_nanos/]
// TESTRESPONSE[s/"score": 4552/"score": $body.profile.shards.0.searches.0.query.0.children.1.breakdown.score/]
// TESTRESPONSE[s/"build_scorer": 42602/"build_scorer": $body.profile.shards.0.searches.0.query.0.children.1.breakdown.build_scorer/]
// TESTRESPONSE[s/"create_weight": 89323/"create_weight": $body.profile.shards.0.searches.0.query.0.children.1.breakdown.create_weight/]
// TESTRESPONSE[s/"next_doc": 2852/"next_doc": $body.profile.shards.0.searches.0.query.0.children.1.breakdown.next_doc/]
// TESTRESPONSE[s/"time_in_nanos": "304311"/"time_in_nanos": $body.profile.shards.0.searches.0.collector.0.time_in_nanos/]
// TESTRESPONSE[s/"time_in_nanos": "32273"/"time_in_nanos": $body.profile.shards.0.searches.0.collector.0.children.0.time_in_nanos/]
```

To how you can cavalierly mock all the numbers at once with this change:
```
// TESTRESPONSE[s/(?<=[" ])\d+(\.\d+)?/$body.$_path/]
```

### CODE REVIEW

**Comment by nik9000** [comment]
@colings86 had asked for some way to mock things in the response snippets which would have required recursive unstashing - the stash keys contained stash keys that had to be resolved. I tried this out and it became pretty silly pretty fast. So instead I put together this which is a thing I've been thinking about on and off for a while. I think it'll allow the kinds of mocking @colings86 needs and it could make writing a lot of the response snippets easier. I applied this to the profile API snippets because those always made me sad. They aren't *easy* to read now, but they are much easier.

**Comment by polyfractal** [comment]
++ the sentiment, I've run into similar issues.  The syntax does confuse me a bit though.  E.g.

```
// TESTRESPONSE[s/"hits": \[...\]/"hits": $body.$_path/]
```

So I know that's replacing `"hits": [...]` with whatever the `hits` were in the response, but I don't see how `$body.$_path` translates to "take whatever is in the hits"?  

Maybe we could introduce a new kind of operator, like:

```
// TESTSTRUCTURE["$body.hits"]
```

Which only checks the keys but not the values of the path?  It's basically the same thing, but not having to deal with the stashing mechanics and regex replace?  That said, not sure how much difficulty that adds to the code to implement it, I can see the appeal of continuing to piggyback on stash mechanics :)

**Comment by nik9000** [comment]
The trouble with something like `// TESTSTRUCTURE["$body.hits"]` is that it'd be had to do selectively. In my experience "take whatever is in the response" is something you want to do only for certain parts of the response. Even for the profiling API I only want to mock the numbers and not the `type` or `description` fields. Basically, doing it at the "top level" would require only matching the structure at the top level without a way pull out parts of the lower level that you want to match.

Another way to go about this same thing is instead of comparing the maps with equality we could compare with matchers, but then you'd need a syntax to build the matchers. Which is not a can of worms I wanted to open up.

The syntax only really makes sense from an "unstashing" perspective. I could add an abstraction on top of it to make it shorter but I'm not sure I'm weary of the number of abstractions I'm adding here. We're already compiling snippets to yaml tests by way of regex replacements.

I dunno, I read `"$body.$_path" as "copy whatever is at the current path from the body".

**Comment by polyfractal** [comment]
Yeah, that's a good point.  Hadn't thought about wanting to mix/match real tests with "just fill in the blanks please".

:+1:  from me :)

**Comment by nik9000** [comment]
> 👍 from me :)

If you have a suggestion on syntax or naming I'd really appreciate it. Or to rewrite the docs I added for it. Or something like that. I'm fully aware that the name isn't great.

**Comment by polyfractal** [comment]
I can't think of a better alternative.  I agree it's a bit clunky, but I don't have any better suggestions  ¯\\\_(ツ)\_/¯

I think adding a slightly more common example to the docs would be good though.  E.g. showing how `"id": "$body.profile.shards.0.id"` could be rewritten to `"id": "$body.$_path`, that way the user knows the sort of situation that it's useful for.  The took example is a bit contrived since it doesn't really save you much hassle :)

**Review by colings86 — APPROVED** [review]
This LGTM but I don't know this code particularly well so maybe @javanna should also look before we merge?

File: docs/README.asciidoc [hunk 1]
```diff
@@ -47,6 +47,15 @@ for its modifiers:
   how it works. These are much more common than `// TEST[s/foo/bar]` because
   they are useful for eliding portions of the response that are not pertinent
   to the documentation.
+    * One interesting difference here is that you often want to match against
+    the response from Elasticsearch. To do that you can reference the "body" of
+    the response like this: `// TESTRESPONSE[s/"took": 25/"took": $body.took/]`.
+    Note the `$body` string. This says "I don't expect that 25 number in the
+    response, instead just match against what is in the response." Instead of
+    writing the path into the response after `$body` you can instead write
```
Comment (javanna) [inline docs/README.asciidoc]: "instead" is repeated twice in this sentence.

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/Features.java [hunk 2]
```diff
@@ -41,6 +41,7 @@
             "headers",
             "stash_in_key",
             "stash_in_path",
+            "stash_path_magic_key",
```
Comment (javanna) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/Features.java]: as much as I like magic, I would love to find a better name for this. stash_path_replace or something like that?

**Review by javanna — APPROVED** [review]
left a couple of minors, LGTM otherwise

File: test/framework/src/main/java/org/elasticsearch/test/rest/yaml/Features.java (hunk 2 above)
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/rest/yaml/Features.java]: `stash_path_replace` is great. I'll use that.

**Comment by nik9000** [comment]
Thanks for all the reviews! I've merged this to master and backported to 5.x.

**Pull request merged**